packages/hash/frontend/src/lib/entities.ts

```diff
@@ -1,127 +1,419 @@
- * - adds the contents of its 'properties' field to the root
- * - delete the 'properties' field
- * @returns void
+ * Converts an entity from its GraphQL API representation to its Block Protocol representation:
+ * 1. Only provide 'entityId' and 'entityTypeId' at the top level
+ * 2. Re-write 'entityId' so that it is a stringified object of the identifiers we need (i.e. to include accountId)
  */
-const destructivelyMoveEntityPropertiesToRoot = (
-  entity: Partial<UnknownEntity>,
-  preserveExtraMetadata: boolean = false,
-) => {
-  if (!preserveExtraMetadata) {
-    for (const key of Object.keys(entity)) {
-      if (
-        key !== "id" &&
-        key !== "accountId" &&
-        key !== "entityId" &&
-        key !== "entityType" &&
-        key !== "linkGroups" &&
-        key !== "linkedEntities" &&
-        key !== "linkedAggregations" &&
-        key !== "properties"
-      ) {
-        delete entity[key as keyof UnknownEntity];
-      }
-    }
-  }
-  Object.assign(entity, entity.properties);
-  delete entity.properties;
+export const convertApiEntityToBpEntity = ({
+  accountId,
+  entityId,
+  entityTypeId,
+  properties,
+}: Pick<
+  ApiEntity,
+  "accountId" | "entityId" | "entityTypeId" | "properties"
+>): BpEntity => {
+  if (entityId.includes("{")) {
+    throw new Error(
+      `entityId has already been re-written as a stringified object: ${entityId}`,
+    );
+  }
+  return {
+    entityId: rewriteEntityIdentifier({ accountId, entityId }),
+    entityTypeId,
+    properties,
+  };
 };
 
 /**
- * Creates a deep clone of an entity with the contents of the 'properties'
- * field moved to the root of the cloned object.
+ * Converts entities from their GraphQL API representation to their Block Protocol representation.
+ * @see convertApiEntityToBpEntity
+ */
+export const convertApiEntitiesToBpEntities = (
+  records: Pick<
+    ApiEntity,
+    "accountId" | "entityId" | "entityTypeId" | "properties"
+  >[],
+): BpEntity[] => records.map(convertApiEntityToBpEntity);
+
+/**
+ * We send blocks an 'entityId' that is a stringified object in {@link convertApiEntityToBpEntity}
+ * – this reverses the process so we have the individual fields to use in calling the HASH API.
  *
- * Deletes existing root fields which aren't related to the entity's id or type,
- * UNLESS false is passed as the second argument.
- * @returns the cloned entity, transformed
+ * @param stringifiedIdentifier any 'entityId' or equivalent (e.g. sourceEntityId) sent from a block
  */
-const cloneEntityWithPropertiesAtRoot = (
-  entity: Partial<UnknownEntity>,
-  preserveExtraMetadata: boolean = false,
-) => {
-  const clone = cloneDeep(entity);
-  destructivelyMoveEntityPropertiesToRoot(clone, preserveExtraMetadata);
-  return clone;
+export function parseEntityIdentifier(
+  stringifiedIdentifier: string,
+): ApiEntityIdentifier {
+  let identifierObject: ApiEntityIdentifier;
+  try {
+    identifierObject = JSON.parse(stringifiedIdentifier);
```

Comment:
So this doesn't look like you're necessarily doing anything wrong. The basic example errors on the playground. Will look into this. 

<img width="913" alt="image" src="https://user-images.githubusercontent.com/719814/178334549-d2605f39-65cf-4324-b5d5-ca74f4311943.png">

<img width="413" alt="image" src="https://user-images.githubusercontent.com/719814/178334555-164dec14-93a7-467c-9bf0-eb98864961f8.png">

Reply:
The explanation / solution is at https://stackoverflow.com/a/70029241/851985 and the Github issue is at https://github.com/microsoft/TypeScript/issues/21732

> This is currently a missing feature in TypeScript. See [microsoft/TypeScript#21732](https://github.com/microsoft/TypeScript/issues/21732) for the relevant feature request. Right now when you use the [in operator as a type guard](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) like k in result, it only really does any meaningful narrowing if the type of result is a [union](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types) where some members are known to have the key k and others are not. It does not assert the existence of a property with key k. So in your case, where the type of result is just object (narrowed from unknown), nothing meaningful happens.
> 
> The workaround I sometimes use is to write my own [user-defined type guard function](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) which behaves the way I want in to behave. For example:
> 
> function hasKey<K extends string, T extends object>(
>     k: K, o: T
> ): o is T & Record<K, unknown> {
>     return k in o;
> }
> Now, instead of writing k in result, I write hasKey(k, result):
> 
> if (
>     typeof result === "object" &&
>     result !== null &&
>     hasKey("movieName", result) &&
>     typeof result.movieName === "string" // okay
> ) {
>     result.movieName.toUpperCase(); // okay
> }
> This works how you want, and is arguably the closest you can get to type safety here, at least until something is done to address microsoft/TypeScript#21732 natively.

Reply:
Very helpful, thank you! Now addressed in https://github.com/hashintel/hash/pull/752/commits/c6e4e2e1c5d6b11ec99b44c4f07e1dcad718e020